Fix watch negotiation when using a non-default mime type in the client

smarterclayton opened the pull request:
/kind bug

Fix negotiation of watch protocols in the client by removing a lot of hardcoded logic and switching from NegotiatedSerializer (returns a list of supported protocols) to ClientNegotiator (which does negotiation).  Also simplifies and removes unnecessary conversion logic within the client, leaving that part of the negotiator.  After this change a user should be able to request a watch for protobuf and json and get a json watch for CRDs, which means clients can simply describe what protocols they support and get the correct behavior (without having to hardcode use of JSON for CRDs).

This goes a bit further than #62175 and makes some other changes to RESTClient that have been delayed for a while, specifically adding an interface specific to client-> server interactions.  That makes it easier to remove conversion specific logic from rest client (which is long overdue) behind that interface, as well as setting the stage for a single client to handle regular Kube objects in Protobuf but CRD objects in JSON (so we don't have to have two clients).  For now, focus on streamlining how RESTClient converts objects to bytes and back.

The RESTClient and Request objects now have much less overlap - Requests take a RESTClient and only track local state relevant to the request.  Access to private members (throttle/ratelimiter) is now restricted. Notable test changes and isolated refactors are pulled out into the preceding commits (which we can merge here or in separate PRs as necessary).  The last commit changes the RESTClient interface.

The e2e test reproduces #62175 and will be expanded to check for other negotiation related errors against real servers.

```release-note
Clients can request protobuf and json and correctly negotiate with the server for JSON for CRD objects, allowing all client libraries to request protobuf if it is available.  If an error occurs negotiating a watch with the server, the error is immediately return by the client `Watch()` method instead of being sent as an `Error` event on the watch stream.
```

smarterclayton committed:
Detect watch protocol errors via an e2e test for apimachinery

This e2e test reproduces #62175 and will be expanded to check for other
negotiation related errors against real servers.

smarterclayton committed:
Create a shim for Codecs that handles client duties

Clients have to renegotiate frequently, and the shape of the interface
that a client wants is slightly different than the interface on the
server. Create a new ClientNegotiator interface that represents what
the client->server code would want to use (mostly, still evolving
Encoder) and move versioning details out of RESTClient.

In the long run, we want to remove internal clients and conversions
from clients. This takes a step in that direction and also makes sure
watch negotiation is consistent with the server.

smarterclayton committed:
Remove deprecated-dynamic client

It is now unused.

smarterclayton committed:
test: Watch should fail immediately on negotiate errors

Instead of returning an error on the watch stream, if we can't properly
negotiate a watch serialization format we should error and return that
error to the client.

smarterclayton committed:
dynamic: The dynamic client no longer needs a special cased watch

By correctly handling content type negotiation, we can avoid the
need for a special version of watch and use the same code path as
typed clients.

smarterclayton committed:
test: Set RateLimiter via client config vs direct casting

This test has no need to perform a dereference to an implementation,
it can instead set the rate limited during initialization.

smarterclayton committed:
test: kubectl unit tests should be using codecs without conversion

Tests are also refactored to use the simpler RESTClient code path.

smarterclayton committed:
test: Exit early during resource helper test

The test will panic if it fails, and should instead check and exit
when invalid conditions are hit.

smarterclayton commented:
/assign @liggitt 

This is ready for review, refreshed and should be passing (although haven't completely 

@kubernetes/sig-api-machinery-pr-reviews

smarterclayton commented:
/priority important-soon

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -98,50 +99,79 @@ type Request struct {
+	})
+}
+
+// NewRequestWithClient is exposed to allow test interfaces to create a request with a different RESTClient.
```
this comment is confusing... isn't it still constructing the same RESTClient in the method?

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -595,6 +614,7 @@ func (r *Request) WatchWithSpecificDecoders(wrapperDecoderFn func(io.ReadCloser)
 		if net.IsProbableEOF(err) {
 			return watch.NewEmptyWatch(), nil
 		}
+		fmt.Printf("DEBUG: %T %#v\n", err, err)
```
drop debug

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -604,9 +624,22 @@ func (r *Request) WatchWithSpecificDecoders(wrapperDecoderFn func(io.ReadCloser)
+	}
+
+	frameReader := framer.NewFrameReader(resp.Body)
+	streamingDecoder := streaming.NewDecoder(frameReader, streamingSerializer)
```
is it unreasonable of me to be confused that `StreamDecoder()` returns `decoder`, from which we construct `streamingDecoder`, and then we pass both into `NewDecoder(...)`? can we name these better?

smarterclayton replied:
I mean, we can.  They're the same names we use today.

liggitt replied:
would these names be clearer (and accurately match use)?
* streamingDecoder -> watchEventDecoder
* decoder -> objectDecoder

liggitt commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -56,24 +56,30 @@ import (
 
 func TestNewRequestSetsAccept(t *testing.T) {
-	r := NewRequest(nil, "get", &url.URL{Path: "/path/"}, "", ContentConfig{}, Serializers{}, nil, nil, 0)
+	r := NewRequestWithoutClient(&url.URL{Path: "/path/"}, "", ClientContentConfig{}).Verb("get")
```
does case matter for verb? "POST" below, but "get" here

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/attach/attach_test.go:
```diff
@@ -239,7 +239,7 @@ func TestAttach(t *testing.T) {
 
 			codec := scheme.Codecs.LegacyCodec(scheme.Scheme.PrioritizedVersionsAllGroups()...)
-			ns := scheme.Codecs
+			ns := scheme.Codecs.WithoutConversion()
```
how do we know to sweep everywhere we need to switch to `WithoutConversion()`?

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_clusterrolebinding_test.go:
```diff
@@ -79,6 +78,7 @@ func TestCreateClusterRoleBinding(t *testing.T) {
 	tf.Client = &ClusterRoleBindingRESTClient{
 		RESTClient: &fake.RESTClient{
+			GroupVersion:         schema.GroupVersion{Group: "rbac.authorization.k8s.io", Version: "v1beta1"},
 			NegotiatedSerializer: ns,
```
should `ns` above switch to `WithoutConversion()`?

smarterclayton replied:
Will switch all of these tests to use it.

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_rolebinding_test.go:
```diff
@@ -81,6 +80,7 @@ func TestCreateRoleBinding(t *testing.T) {
 	tf.Client = &RoleBindingRESTClient{
 		RESTClient: &fake.RESTClient{
+			GroupVersion:         groupVersion,
 			NegotiatedSerializer: ns,
```
WithoutConversion()?

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -155,6 +155,15 @@ type NegotiatedSerializer interface {
+// Use NewClientNegotiator or NewVersionedClientNegotiator to create this interface from
+// a NegiotatedSerializer.
+type ClientNegotiator interface {
+	Encoder(contentType string, params map[string]string) (Encoder, error)
```
Should be simple (params is the charset).  That's technically the naming that the RFC uses, will clarify.

It should return a negotiation (content type not acceptable) error.  Will clarify.

smarterclayton replied:
That's handled in `NegotiateOutputMediaType`.

smarterclayton replied:
Added an Error type

liggitt replied:
> optional mediaType parameters

does that mean all parameters must be considered optional? how does that reconcile with our use of them to describe alternate GVKs?

liggitt replied:
would be good to describe supported params or behavior if unknown params are passed

smarterclayton replied:
Servers don't return media type parameters today (will double check).  It's likely if they did other clients would break.  I think the best we can do is to say "it is up to the client negotiator whether to treat these as optional or required, but current implementations consider them optional and modifiers to the base content type.

liggitt replied:
not super satisfying, but after digging through I agree it's not worse than today, and it's not clear we could safely start returning parameters from the server

smarterclayton commented on staging/src/k8s.io/client-go/rest/fake/fake.go:
```diff
@@ -85,28 +85,16 @@ func (c *RESTClient) GetRateLimiter() flowcontrol.RateLimiter {
 }
 
-func (c *RESTClient) Do(req *http.Request) (*http.Response, error) {
+func (c *RESTClient) do(req *http.Request) (*http.Response, error) {
```
Yes, so that there is no confusion that RESTClient does not magically implement rest.HTTPClient based on the construction.

smarterclayton replied:
Added a lot more doc to fake.RESTClient and cleaned up the interface a bit more.

smarterclayton commented on staging/src/k8s.io/client-go/rest/fake/fake.go:
```diff
@@ -85,28 +85,16 @@ func (c *RESTClient) GetRateLimiter() flowcontrol.RateLimiter {
-		serializers.Framer = info.StreamSerializer.Framer
-	}
-	return restclient.NewRequest(c, verb, &url.URL{Host: "localhost"}, c.VersionedAPIPath, config, serializers, nil, nil, 0)
+	return restclient.NewRequestWithClient(&url.URL{Scheme: "https", Host: "localhost"}, c.VersionedAPIPath, config, CreateHTTPClient(c.do))
```
RESTClient needs a client to do anything.  Before, that was done with clever faking (the public `Do`) which lead to code not being clear about whether it used RESTClient as an http.RoundTripper.

That is now corrected and the ONLY client passed to rest.Request is an http.Client, it must ALWAYS be specified, and for the fake it ALWAYS has the http.Client with a round tripper that calls `do()` on the fake.RESTClient.

smarterclayton replied:
Added a lot more doc to fake.RESTClient and cleaned up the interface a bit more.

smarterclayton commented on staging/src/k8s.io/client-go/rest/client.go:
```diff
@@ -64,66 +84,42 @@ type RESTClient struct {
-// decoding of responses from the server.
-func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ContentConfig, maxQPS float32, maxBurst int, rateLimiter flowcontrol.RateLimiter, client *http.Client) (*RESTClient, error) {
+// such as Get, Put, Post, and Delete on specified paths.
+func NewRESTClient(baseURL *url.URL, versionedAPIPath string, config ClientContentConfig, rateLimiter flowcontrol.RateLimiter, client *http.Client) (*RESTClient, error) {
```
Yeah, I refrained from fixing that.  RESTClients don't assume group version matches versionedAPIPath today, since construction of versionedAPIPath can depend on the group in some of the existing code (core).

smarterclayton replied:
Also because RESTClient exposes `GroupVersion()` and `Request#VersionedParams` needs it.

I'll add a comment to ClientContentConfig to that effect.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/endpoints/watch_test.go:
```diff
@@ -659,38 +658,20 @@ func TestWatchHTTPDynamicClientErrors(t *testing.T) {
 	}
-	if status.Kind != "Status" || status.APIVersion != "v1" || status.Code != 500 || status.Status != "Failure" || !strings.Contains(status.Message, "we got an error") {
-		t.Fatalf("error: %#v", status)
+	if err.Error() != "no stream serializers registered for testcase/json" {
```
This is strictly better - we are now in the client library, as part of watch establishment, able to detect that whatever the client gave us is invalid.  This means you fail immediately upon watch establishment if you asked for something impossible, and the error is not eaten (this is really the entire change the PR is trying to result in - sane negotiation at the time the request is dispatched).  The commit containing only this test change has more description.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/runtime/negotiate.go:
```diff
@@ -0,0 +1,130 @@
+	encode, decode GroupVersioner
+}
+
+func (n *clientNegotiator) Encoder(contentType string, params map[string]string) (Encoder, error) {
```
It might mean `pretty` is broken.  I'll double check.

smarterclayton replied:
Pretty is handled in `NegotiateOutputMediaType`, none of the code that uses the client negotiator now previously handled Pretty.  I think that can be a follow up cleanup.

smarterclayton replied:
Added a TODO

liggitt replied:
it still feels broken to ignore the params... doesn't that mean it is really hard to use that to modify encoding/decoding in the future compatibly? Do we have scenarios that are populating these today? An initial implementation that returned an error if unknown params were required would make more sense to me (both for Encode/Decode/StreamDecoder)

smarterclayton replied:
This is no different than the old code.  Client previously ignored these during negotiation.  In a sense these are informational - the only behavior we key off of today is pretty (at a higher level).

If we did add future behavior here, it would be backward compatible because this is client focused and already has to depend on the values in serializer info.  If they were populated by a server (hypothetically) we'd completely ignore them on the way back, and then a client that was smarter would have to start handling them.  If we started failing in the client now on unrecognized options, it's possible that we'd be broken against one or more server implementations (although very unlikely).

liggitt replied:
> If we started failing in the client now on unrecognized options, it's possible that we'd be broken against one or more server implementations

Yeah, that's what I was wondering (like a LB that rewrote to `Content-Type: application/json;charset=utf8` or something supposedly no-op)

smarterclayton commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -310,7 +305,31 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
+		if config.Burst == 0 {
+			burst = DefaultBurst
+		}
+		rateLimiter = flowcontrol.NewTokenBucketRateLimiter(qps, burst)
```
Yeah, my bad.  Also we don't have a test for that.  Will add.

smarterclayton commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -98,50 +99,79 @@ type Request struct {
-	throttle   flowcontrol.RateLimiter
 }
 
 // NewRequest creates a new request helper object for accessing runtime.Objects on a server.
```
I don't think we ever allowed mutating a RESTClient directly once created.

smarterclayton replied:
Client is public because a bunch of old testing code still directly copies the Client from a fake to the test rest client.

smarterclayton commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -98,50 +99,79 @@ type Request struct {
 	return r
 }
 
+// NewRequestWithoutClient is exposed to allow test interfaces to create a request with a different RESTClient.
```
Collapsing everything to NewRequestWithClient which will clearly be for test code.

smarterclayton commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -397,14 +427,19 @@ func (r *Request) Body(obj interface{}) *Request {
 			return r
 		}
-		data, err := runtime.Encode(r.serializers.Encoder, t)
+		encoder, err := r.c.content.Negotiator.Encoder(r.c.content.ContentType, nil)
```
Technically `application/json` is a content type and `application/json;pretty=1` is a mediaTypeWithParams.  We don't let you get Pretty encoding today.

liggitt replied:
so setting client.ClientContentConfig.ContentType = "application/json;as=..." would be invalid?

smarterclayton commented on test/e2e/network/service_latency.go:
```diff
@@ -77,9 +77,12 @@ var _ = SIGDescribe("Service endpoints latency", func() {
+			framework.Failf("Unable to load config: %v", err)
+		}
+		cfg.RateLimiter = flowcontrol.NewFakeAlwaysRateLimiter()
+		f.ClientSet = kubernetes.NewForConfigOrDie(cfg)
```
We're just creating a new ClientSet for the duration of this test.  The framework lives for exactly one test, then ClientSet is cleared at the end of a test (or was, when I wrote this)

smarterclayton commented on test/e2e/apimachinery/protocol.go:
```diff
@@ -0,0 +1,80 @@
+	f := framework.NewDefaultFramework("protocol")
+	f.SkipNamespaceCreation = true
+
+	for _, s := range []string{
```
Would need to be a new test, you can't do that on Typed clients (they cast to Service).

smarterclayton commented:
review comments in a new commit, will squash it in at the end

liggitt commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1000,63 +1024,72 @@ func TestRequestWatch(t *testing.T) {
 			},
 			Empty: true,
 		},
+		// After changing to a transport, we already wrap io.EOF with a url.Error
```
drop commented code?

liggitt commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1826,6 +1927,46 @@ func TestWatch(t *testing.T) {
+	}))
+	defer testServer.Close()
+
+	// set the default content type to protobuf so that we test falling back to JSON serialization
```
copy/pasta comment... this doesn't fall back to JSON

liggitt commented on staging/src/k8s.io/client-go/rest/fake/fake.go:
```diff
@@ -41,40 +43,49 @@ func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Req will not be recorded, Resp will not be returned, and Client will not be
+	// invoked.
+	Err error
+	// Req is set to the client that this RESTClient creates.
```
to the client?

smarterclayton commented:
> so setting client.ClientContentConfig.ContentType = "application/json;as=..." would be invalid?

Correct

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -604,9 +615,22 @@ func (r *Request) WatchWithSpecificDecoders(wrapperDecoderFn func(io.ReadCloser)
 	}
-	wrapperDecoder := wrapperDecoderFn(resp.Body)
+
+	contentType := resp.Header.Get("Content-Type")
```
This content type will have `;stream=watch` for proto watch streams, which we'll parse to params, and ignore in the StreamDecoder impl. Just pointing that out.

liggitt commented:
this lgtm, feel free to squash

smarterclayton committed:
Always negotiate a decoder using ClientNegotiator

This commit performs two refactors and fixes a bug.

Refactor 1 changes the signature of Request to take a RESTClient, which
removes the extra copy of everything on RESTClient from Request. A pair
of optional constructors are added for testing. The major functional
change is that Request no longer has the shim HTTPClient interface and
so some test cases change slightly because we are now going through
http.Client code paths instead of direct to our test stubs.

Refactor 2 changes the signature of RESTClient to take a
ClientContentConfig instead of ContentConfig - the primary difference
being that ClientContentConfig uses ClientNegotiator instead of
NegotiatedSerializer and the old Serializers type. We also collapse
some redundancies (like the rate limiter can be created outside the
constructor).

The bug fix is to negotiate the streaming content type on a Watch()
like we do for requests. We stop caching the decoder and simply
resolve it on the request. We also clean up the dynamic client
and remove the extra WatchSpecificVersions() method by providing
a properly wrapped dynamic client.

smarterclayton commented:
Squashed to minimal commits

smarterclayton commented:
Will tag unless there are more comments

dims commented on staging/src/k8s.io/apimachinery/pkg/runtime/negotiate.go:
```diff
@@ -0,0 +1,159 @@
+	return fmt.Sprintf("no serializers registered for %s", e.ContentType)
+}
+
+func NewCodecForContentTypeOrDie(ns NegotiatedSerializer, contentType string, gv GroupVersioner) Codec {
```
Is `NewCodecForContentTypeOrDie` used anywhere?

smarterclayton replied:
No, I was going to use it but it's not critical right now.  Removing.

liggitt commented:
/lgtm

liggitt commented:
/hold for dims' question

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest
